protos/index.proto

```diff
@@ -22,25 +22,65 @@ package lance.index.pb;
 
 // The type of an index.
 enum IndexType {
-NONE = 0;
+  NONE = 0;
 
-// Flat index
-VECTOR_FLAT = 100;
+  // Flat index
+  VECTOR_FLAT = 100;
 
-// IVF + PQ index
-VECTOR_IVF_PQ = 101;
+  // IVF + PQ index
+  VECTOR_IVF_PQ = 101;
 }
 
 message Index {
-    // The unique index name in the dataset.
-    string name = 1;
+  // The unique index name in the dataset.
+  string name = 1;
 
-    // Columns to be used to build the columns.
-    repeated string columns = 2;
+  // Columns to be used to build the columns.
+  repeated string columns = 2;
 
-    // The version of the dataset this index was built from.
-    uint64 dataset_version = 3;
+  // The version of the dataset this index was built from.
+  uint64 dataset_version = 3;
 
-    // The [`IndexType`] of the index.
-    IndexType index_type = 4;
+  // The [`IndexType`] of the index.
+  IndexType index_type = 4;
+
+  oneof implementation {
+    VectorIndex vector_index = 5;
+  }
+}
+
+// Inverted Index File Metadata.
+message IVF {
+  // Centroids of parittions. `dimension * num_partitions` of float32s.
+  repeated float centroids = 1;
```

Comment:
> should this be repeated Centroid where Centroid is repeated float

When loading into memory, all centroids are put into one `flatten` array, so it is easy to make it all cache-aligned, and to do SIMD. Which seems not necessary to just keep the 2D array on disk?

> curious why store centroids in pb rather than as arrow fixedsizelist array somehwere?

So it was for saving I/Os,  i assume that usually the average vector index would be 256-1024 dimension of floats, up to 256.  By storing them in the protobuf, we can just use 1 I/O to read the metadata block. Otherwise, we need read metadata block (for reading the position of an array fixedsizelist), and do another I/O to read the array.

And since these centroids are necessary no matter what query is, it is not like the "page table" case in the data layout, which offers a chance to read partial data to satisfy a read.